gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/IcebergMetadataWriter.java

```diff
@@ -0,0 +1,828 @@
+        gobblinTrackingEvent.addMetadata(entry.getKey(), entry.getValue());
+      }
+    }
+    eventSubmitter.submit(gobblinTrackingEvent);
+  }
+
+  private boolean setDatasetOffsetRange(TableMetadata tableMetadata, Map<String, String> props) {
+    if (tableMetadata.dataOffsetRange.isPresent() && !tableMetadata.dataOffsetRange.get().isEmpty()) {
+      for (Map.Entry<String, List<Range>> offsets : tableMetadata.dataOffsetRange.get().entrySet()) {
+        List<Range> ranges = offsets.getValue();
+        String rangeString = ranges.stream()
+            .map(r -> Joiner.on(ConfigurationKeys.RANGE_DELIMITER_KEY).join(r.lowerEndpoint(), r.upperEndpoint()))
+            .collect(Collectors.joining(ConfigurationKeys.LIST_DELIMITER_KEY));
+        props.put(String.format(OFFSET_RANGE_KEY_FORMAT, offsets.getKey()), rangeString);
+      }
+      return true;
+    }
+    return false;
+  }
+
+  private void updateSchema(TableMetadata tableMetadata, Map<String, String> props, String topicName) {
+    //Set default schema versions
+    props.put(SCHEMA_CREATION_TIME_KEY, tableMetadata.lastSchemaVersion.or(DEFAULT_CREATION_TIME));
+    // Update Schema
+    try {
+      if (tableMetadata.candidateSchemas.isPresent() && tableMetadata.candidateSchemas.get().size() > 0) {
+        Cache candidates = tableMetadata.candidateSchemas.get();
+        //Only have default schema, so either we calculate schema from event or the schema does not have creation time, directly update it
+        if (candidates.size() == 1 && candidates.getIfPresent(DEFAULT_CREATION_TIME) != null) {
+          updateSchemaHelper(DEFAULT_CREATION_TIME, (Schema) candidates.getIfPresent(DEFAULT_CREATION_TIME), props,
+              tableMetadata.table.get());
+        } else {
+          //update schema if candidates contains the schema that has the same creation time with the latest schema
+          org.apache.avro.Schema latestSchema =
+              (org.apache.avro.Schema) schemaRegistry.getLatestSchemaByTopic(topicName);
+          String creationTime = AvroUtils.getSchemaCreationTime(latestSchema);
+          if (creationTime == null) {
+            log.warn(
+                "Schema from schema registry does not contain creation time, check config for schema registry class");
+          } else if (candidates.getIfPresent(creationTime) != null) {
+            updateSchemaHelper(creationTime, (Schema) candidates.getIfPresent(creationTime), props,
+                tableMetadata.table.get());
+          }
+        }
+      }
+    } catch (SchemaRegistryException e) {
+      log.error("Cannot get schema form schema registry, will not update this schema", e);
+    }
+  }
+
+  private void updateSchemaHelper(String schemaCreationTime, Schema schema, Map<String, String> props, Table table) {
+    try {
+      props.put(SCHEMA_CREATION_TIME_KEY, schemaCreationTime);
+      table.updateSchema().unionByNameWith(schema).commit();
+    } catch (Exception e) {
```

Comment:
If not trying to be very protective I think we should catch explicit exception as a best practice.

Reply:
This is mainly because the API definition does not specify which type of exception it will throw, and I do want to make sure that all exception thrown by schema update will not fail the job. Since most of the time it's because iceberg does not support such evolution, which should not affect the job process. I will update the code to make sure schema creation key set after the schema change, so in this case, even the schema update does not succeed, next flush it will try again.